### PR TITLE
IAT-2432:  implement keyword search

### PR DESCRIPTION
**Pull request by gregcharles**

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            "      WHERE item.key=item_keyword.item_key\n" +
+            "      AND %s\n" +
+            "      ORDER BY item_keyword.key LIMIT 1)\n";
+
+    private final String value;
+    private final boolean caseSensitive;
+
+    private final Converter<KeywordSearchFilter, String> syntaxConverter;
+
+    @SuppressWarnings("WeakerAccess")
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java]: I'm sure I'm missing something, but you'd get a weaker access warning on this constructor?

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public String getJoinClause() {
+        final String query = syntaxConverter.convert(this);
+        return String.format(joinTemplate, query);
+    }
+
+    @Override
+    public Optional<FilterQuery> getFilterQuery() {
+        throw new NotImplementedException(
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java]: Rather than throw is there any harm in returning an `Optional.empty()`?  Or is there a design change we need to think making where this throw isn't necessary.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchSyntaxConverter.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
+package org.opentestsystem.ap.imrt.iss.model;
+
+//import org.opentestsystem.ap.imrt.iss.util.javaluator.*;
+import com.fathzer.soft.javaluator.*;
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchSyntaxConverter.java]: Please remove the commented import.  Also, in general IMRT doesn't use the `*` import.  This can easily be done in IntelliJ.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java [hunk 4]
```diff
@@ -23,81 +23,7 @@
...
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.TTS_VISUAL_REQUIRED;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.UPDATED_DATE;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.VALIDATION_RESULT_SEVERITY_COUNT_BENIGN;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.VALIDATION_RESULT_SEVERITY_COUNT_DEGRADED;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.VALIDATION_RESULT_SEVERITY_COUNT_SEVERE;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.VALIDATION_RESULT_SEVERITY_COUNT_TOLERABLE;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.WORKFLOW_STATUS;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.WORKFLOW_STATUS_UPDATE_DATE;
-import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.WRITING_PURPOSE;
+import static org.opentestsystem.ap.imrt.iss.dto.search.SearchProperty.*;
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: Same comment about `*`

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java [hunk 5]
```diff
@@ -178,9 +104,12 @@
             VALIDATION_RESULT_SEVERITY_COUNT_DEGRADED.getColumnName() + ", \n" +
             VALIDATION_RESULT_SEVERITY_COUNT_TOLERABLE.getColumnName() + ", \n" +
             VALIDATION_RESULT_SEVERITY_COUNT_BENIGN.getColumnName() + ", \n" +
+//            KEYWORD_CONTENT.getColumnName() + ", \n" +
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: please remove

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java [hunk 6]
```diff
@@ -285,7 +214,9 @@ public int getTotalItemsCount(final List<SearchFilter> searchFilters) {
                     .withSevereValidationResultCount(String.valueOf(rs.getInt(VALIDATION_RESULT_SEVERITY_COUNT_SEVERE.getColumnName())))
                     .withDegradedValidationResultCount(String.valueOf(rs.getInt(VALIDATION_RESULT_SEVERITY_COUNT_DEGRADED.getColumnName())))
                     .withTolerableValidationResultCount(String.valueOf(rs.getInt(VALIDATION_RESULT_SEVERITY_COUNT_TOLERABLE.getColumnName())))
-                    .withBenignValidationResultCount(String.valueOf(rs.getInt(VALIDATION_RESULT_SEVERITY_COUNT_BENIGN.getColumnName())));
+                    .withBenignValidationResultCount(String.valueOf(rs.getInt(VALIDATION_RESULT_SEVERITY_COUNT_BENIGN.getColumnName())))
+//                    .withKeywordContent(mapToString(rs, KEYWORD_CONTENT.getColumnName()))
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: please remove

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java [hunk 7]
```diff
@@ -92,10 +87,20 @@ private SearchQuery buildQuery(final Optional<PageDto> maybePageDto, final Colle
         Set<String> tables = new HashSet<>();
         MapSqlParameterSource sqlParameterSource = new MapSqlParameterSource();
 
-        if (!searchFilters.isEmpty()) {
-            sb.append(" WHERE \n ");
+        List<SearchFilter> localSearchFilters = new ArrayList<>(searchFilters);
+
+        Optional<KeywordSearchFilter> keywordSearchFilter = removeKeywordSearchFilter(localSearchFilters);
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java]: So if I'm understanding the block below correctly we will always join against item_keyword even if there isn't any filter going against keyword search?  If so, why is that necessary?  I think we should only be joining to tables when we need to join unless the requirement from IAT is to always return a `Section` regardless if we do a keyword search.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java [hunk 8]
```diff
@@ -92,10 +87,20 @@ private SearchQuery buildQuery(final Optional<PageDto> maybePageDto, final Colle
...
+        if (keywordSearchFilter.isPresent()) {
+            sb.append(keywordSearchFilter.get().getJoinClause());
+        } else {
+            sb.append(KeywordSearchFilter.getPlaceholderJoinClause());
+        }
+
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java]: Did you find better performance in having the `WHERE` logic to be within the join rather than in the `WHERE` clause itself?  I realize the join itself is necessary when doing a keyword search.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchQuery.java [hunk 9]
```diff
@@ -5,7 +5,7 @@
 import java.util.Set;
 
 public class SearchQuery {
-    private final String searchQuery;
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchQuery.java]: Why did this need to change?  Is an extending class modifying this value?

File: src/main/resources/application.yml [hunk 10]
```diff
@@ -30,3 +30,8 @@ spring:
   profiles: local-development
 logging:
   config: classpath:log4j2-local.xml
+---
```
Comment (millerts) [inline src/main/resources/application.yml]: Please remove as we're not using zombo.  I understand this was part of your local dev

**Comment by millerts** [comment]
I've made the formatting changes and am merging this PR so QA can move forward testing.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java (hunk 1 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java]: Yes, it's only called by the SearchConverter (and the JUnits), so technically it could be package private. I don''t like using package private though.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java (hunk 2 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchFilter.java]: I changed it to return Optional.empty(). I'll think about the design change.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchSyntaxConverter.java (hunk 3 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/model/KeywordSearchSyntaxConverter.java]: done

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java (hunk 4 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: done

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java (hunk 5 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: Ha, busted! I'm hoping to get this back in later today, so the placeholder will help me, but I've taken it out of this branch now.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java (hunk 6 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/GeneralSearchRepositoryImpl.java]: same

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java (hunk 7 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java]: Discussed offline. If not doing a keyword search, we do a placeholder join to make dummy fields (keyword content and section) with blank strings available in the result set. This eases processing the result set.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java (hunk 8 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchFilterQueryBuilder.java]: It was actually the only way to make the all rows of item_keyword searched for a match, but only return the first one for each item key in the results.

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchQuery.java (hunk 9 above)
Comment (gregcharles) [reply src/main/java/org/opentestsystem/ap/imrt/iss/repository/SearchQuery.java]: Agreed, that's weird. I don't know why I did that. Reverted.

File: src/main/resources/application.yml (hunk 10 above)
Comment (gregcharles) [reply src/main/resources/application.yml]: done
